lean4.html.markdown

```diff
@@ -0,0 +1,587 @@
+-/
+open Group
+infixl:70 " * " => op
+
+/-
+`G` will always stand for a group. Variables `a b c` will be elements of that
+group.
+-/
+variable [Group G] {a b c : G}
+
+def is_identity (e' : G) := ∀ a : G, (a * e' = a ∧ e' * a = a)
+
+/-
+We prove that the identity element is unique.
+-/
+theorem identity_element_unique : ∀ e' : G, is_identity e' → e' = e := by
+  intro e'
+  intro h
+  specialize h e
+  obtain ⟨h1, _⟩ := h
+  have h' := identity e'
+  obtain ⟨_, h2⟩ := h'
+  apply Eq.trans (Eq.symm h2) h1
+/-
+Note that we used the `identity` axiom.
+-/
+
+/-
+Left cancellation. We have to use both `identity` and `inverse` axioms from
+`Group`.
+-/
+theorem left_cancellation : ∀ x y : G, a * x = a * y → x = y := by
+  have h1 := inverse a
+  obtain ⟨ai, a_inv⟩ := h1
+  obtain ⟨_, h2⟩ := a_inv
+  intro x y
+  intro h3
+  calc
+    x = (e : G) * x  := Eq.symm (identity x).right
+    _ = ai * a * x   := by rw [h2]
+    _ = ai * (a * x) := by rw [assoc]
+    _ = ai * (a * y) := by rw [h3]
+    _ = ai * a * y   := by rw [← assoc]
+    _ = (e : G) * y  := by rw [h2]
+    _ = y            := (identity y).right
+
+/-
+Mutually recursive definitions have to be carefully constructed. In particular,
+the hypotheses when recursing should imply well-foundedness.
+
+The game of Nim with two heaps.
+-/
+abbrev between (lower what upper : Nat) : Prop := lower ≤ what ∧ what ≤ upper
+
+mutual
+  def Alice : Nat → Nat → Prop
+    | n1, n2 =>
+      ∃ k, (between 1 k n1 ∧ between 1 k n1 → Bob (n1-k) n2)
+         ∨ (between 1 k n2 ∧ between 1 k n2 → Bob n1 (n2-k))
+
+  def Bob : Nat → Nat → Prop
+    | n1, n2 =>
+      ∀ k, (between 1 k n1 → Alice (n1-k) n2)
+         ∧ (between 1 k n2 → Alice n1 (n2-k))
+end
+
+example : Bob 0 0 := by
+  intro k
+  induction k
+  case zero =>
+    constructor
+    intro ; contradiction
+    intro ; contradiction
+  case succ =>
+    constructor
+    intro a ; have := a.right ; contradiction
+    intro a ; have := a.right ; contradiction
+
+/-
+Programming algorithms in Lean is harder.
+
+This is because we have to convince Lean of termination. Here's a simple
+primality checking algorithm that tests all candidate divisors.
+-/
+def prime' (n : Nat) : Bool :=
+  if h : n < 2 then
+    false
+  else
+    @go 2 n (by omega)
+where
+  go (d : Nat) (n : Nat) {_ : n ≥ d} : Bool :=
+    if h : n = d then /- `h` needed for `omega` below. -/
+      true
+    else if n % d = 0 then
+      false
+    else
+      @go (Nat.succ d) n (by omega)
+  termination_by (n - d)
+/-
+We have to specify that the recursive function `go` terminates because `n-k`
+decreases in each recursive call. This needs the hypothesis `n > k` at the
+recursive call site. But the function itself can only assume that `n ≥ k`. We
+label the test `n ≤ k` by `h` so that the falsification of this proposition can
+be used by `omega` later to conclude that `n > k`.
+
+The tactic `omega` is a general-purpose solver for equalities and inequalities.
```

Comment:
This is also not true. `omega` is a specific-purpose decision procedure for deciding problems in presburger arithmetic.